### PR TITLE
EnC: Log deltas and messages to disk

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 1]
```diff
@@ -1000,7 +1024,11 @@ public async ValueTask<SolutionUpdate> EmitSolutionUpdateAsync(Solution solution
                                 out var moduleNonRemappableRegions,
                                 out var exceptionRegionUpdates);
 
+<<<<<<< HEAD
                             deltas.Add(new ModuleUpdate(
+=======
+                            var delta = new ManagedModuleUpdate(
+>>>>>>> b21317e9e53 (EnC: Log deltas and messages to disk)
```
Comment (davidwengier) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs:1031]: Oops

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs [hunk 2]
```diff
@@ -41,6 +42,25 @@ public EditAndContinueWorkspaceService()
...
+        {
+            try
+            {
+                var path = Environment.GetEnvironmentVariable("Microsoft_CodeAnalysis_EditAndContinue_LogDir");
+                if (string.IsNullOrWhiteSpace(path))
+                {
+                    return null;
+                }
+
+                Directory.CreateDirectory(path);
```
Comment (davidwengier) [inline src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs:55]: Does this throw if the path already exists?

**Review by davidwengier — APPROVED** [review]
I wonder if a boolean env var would be easier... we can find the temp folder, append a datetime to the path maybe? Means you could just leave this it turned on potentially.

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs (hunk 2 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs:55]: It does not.

**Comment by tmat** [comment]
> I wonder if a boolean env var would be easier... we can find the temp folder, append a datetime to the path maybe? Means you could just leave this it turned on potentially.

IDK, sounds like leaving it on would make it easy to create a lot of useless files that way. We'd also need to include VS instance id to avoid multiple VS instances to overwrite each others files.

**Comment by tmat** [comment]
Perhaps we could delete the content of the directory at the start of the first session? That way you can keep it turned on but it will not produced unbounded amount of files.

**Comment by dibarbet** [comment]
> > I wonder if a boolean env var would be easier... we can find the temp folder, append a datetime to the path maybe? Means you could just leave this it turned on potentially.
> 
> IDK, sounds like leaving it on would make it easy to create a lot of useless files by accident. We'd also need to include VS instance id to avoid multiple VS instances to overwrite each others files.

Should this logging use loghub?  Then we don't have to worry about deletion or file size really and they would get uploaded to feedback tickets

**Comment by tmat** [comment]
> Should this logging use loghub? 

Does loghub support logging binary data (potentially large files)?

**Comment by dibarbet** [comment]
Ah good question, not quite sure.  It just gives back a TraceSource
https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/LanguageClient/VisualStudioLogHubLoggerFactory.cs,67

**Comment by tmat** [comment]
I don't think LogHub is suitable for this kind of logging.

**Comment by dibarbet** [comment]
> I don't think LogHub is suitable for this kind of logging.

Sounds good - just wanted to check in case it would be useful!

**Comment by davidwengier** [comment]
Happy to leave it like this for now and see how it goes too :)

**Pull request merged**